samcli/lib/pipeline/bootstrap/stage.py

```diff
@@ -0,0 +1,276 @@
+        """
+        Deploys the CFN template(./stage_resources.yaml) which deploys:
+            * Pipeline IAM User
+            * Pipeline execution IAM role
+            * CloudFormation execution IAM role
+            * Artifacts' S3 Bucket along with KMS encryption key
+            * ECR Repo
+        to the AWS account associated with the given stage. It will not redeploy the stack if already exists.
+        This CFN template accepts the ARNs of the resources as parameters and will not create a resource if already
+        provided, this way we can conditionally create a resource only if the user didn't provide it
+
+        THIS METHOD UPDATES THE STATE OF THE CALLING INSTANCE(self) IT WILL SET THE VALUES OF THE RESOURCES ATTRIBUTES
+
+        Parameters
+        ----------
+        confirm_changeset: bool
+            if set to false, the stage_resources.yaml CFN template will directly be deployed, otherwise, the user will
+            be prompted for confirmation
+
+        Returns True if bootstrapped, otherwise False
+        """
+
+        if self.did_user_provide_all_required_resources():
+            click.secho(f"\nAll required resources for the {self.name} stage exist, skipping creation.", fg="yellow")
+            return True
+
+        missing_resources: str = self._get_non_user_provided_resources_msg()
+        click.echo(f"This will create the following required resources for the {self.name} stage: {missing_resources}")
+        if confirm_changeset:
+            confirmed: bool = click.confirm("Should we proceed with the creation?")
+            if not confirmed:
+                return False
+
+        stage_name_suffix: str = re.sub("[^0-9a-zA-Z]+", "-", self.name)
+        stack_name: str = f"{STACK_NAME_PREFIX}-{stage_name_suffix}-{STAGE_RESOURCES_STACK_NAME_SUFFIX}"
+        stage_resources_template_body = Stage._read_template("stage_resources.yaml")
+        output: StackOutput = manage_stack(
+            stack_name=stack_name,
+            region=self.aws_region,
+            profile=self.aws_profile,
+            template_body=stage_resources_template_body,
+            parameter_overrides={
+                "PipelineUserArn": self.pipeline_user.arn or "",
+                "PipelineExecutionRoleArn": self.pipeline_execution_role.arn or "",
+                "PipelineIpRange": self.pipeline_ip_range or "",
+                "CloudFormationExecutionRoleArn": self.cloudformation_execution_role.arn or "",
+                "ArtifactsBucketArn": self.artifacts_bucket.arn or "",
+                "CreateEcrRepo": "true" if self.create_ecr_repo else "false",
+                "EcrRepoArn": self.ecr_repo.arn or "",
+            },
+        )
+
+        self.pipeline_user.arn = output.get("PipelineUser")
+        self.pipeline_user.access_key_id = output.get("PipelineUserAccessKeyId")
+        self.pipeline_user.secret_access_key = output.get("PipelineUserSecretAccessKey")
```

Comment:
Can we replace the entire `self.pipeline_user` instead of assigning attributes individually? I prefer to make `pipeline_user` an immutable object so that we know it is "updating the entire user object instead of mutating attributes." And it also ensures we never update some of the attributes and forget the others (out of sync)

Reply:
same goes to `self.artifacts_bucket` below

Reply:
We actually update some resources and leave others here; A resource has an attribute `is_user_provided` that is set to true during `init` only if the user provided the ARN during the interactive questionnaire.

Reply:
I mean when we update a resource, we update all attributes of that resource right? If that's the case, doing a single `self.pipeline_user = ...` (with `pipeline_user` being an immutable object like Tuple) is less error prune than having three lines `self.pipeline_user.* = ` to update individual attributes.

Reply:
"when we update a resource, we update all attributes of that resource right?" Nope, we don't, that is the point.